examples/audio/transformer_asr.py

```diff
@@ -0,0 +1,523 @@
+## Callbacks to display predictions and to change learning rate
+"""
+
+
+class DisplayOutputs(keras.callbacks.Callback):
+    def __init__(
+        self, batch, idx_to_token, trg_start_token_idx=27, trg_end_token_idx=28
+    ):
+        """ Displays a batch of outputs after every epoch 
+        Arguments:
+        - batch: test batch containing the keys "src" and "trg"
+        - idx_to_token: a List containing the vocabulary tokens corresponding to their indices
+        - trg_start_token_idx: start token index in the target vocabulary
+        - trg_end_token_idx: end token index in the target vocabulary
+        """
+        self.batch = batch
+        self.trg_start_token_idx = trg_start_token_idx
+        self.trg_end_token_idx = trg_end_token_idx
+        self.idx_to_char = idx_to_token
+
+    def on_epoch_end(self, epoch, logs=None):
+        if epoch % 5 != 0:
+            return
+        src = self.batch["src"]
+        trg = self.batch["trg"].numpy()
+        bs = tf.shape(src)[0]
+        preds = self.model.generate(src, self.trg_start_token_idx)
+        preds = preds.numpy()
+        for i in range(bs):
+            target = ""
+            for idx in trg[i, :]:
+                target += "" + self.idx_to_char[idx]
+            prediction = ""
+            over = False
+            for idx in preds[i, :]:
+                if over:  # Add padding token once end token has beeen predicted
+                    prediction += "-"
+                    continue
+                if idx == self.trg_end_token_idx:
+                    over = True
+                prediction += "" + self.idx_to_char[idx]
+            print(f"target:     {target}")
+            print(f"prediction: {prediction}")
+            print()
+
+
+def scheduler(epoch, lr):
+    """ linear warm up - linear decay """
+    init_lr = 0.00001
+    lr_after_warmup = 0.001
+    final_lr = 0.00001
+    warmup_epochs = 15
+    decay_epochs = 85
+    if epoch < warmup_epochs:
+        return init_lr + ((lr_after_warmup - init_lr) / (warmup_epochs - 1)) * epoch
+    return max(
+        final_lr,
+        lr_after_warmup
+        - (epoch - warmup_epochs) * (lr_after_warmup - final_lr) / (decay_epochs),
+    )
+
+
+"""
+## Create & train the end-to-end model
+"""
+
+
+for i in val_ds.take(1):
+    batch = i  # Use the first batch of validation set to display outputs
+
+# vocabulary to convert predicted indices to characters
+idx_to_char = vectorize_layer.get_vocabulary()
+display_cb = DisplayOutputs(
+    batch, idx_to_char, trg_start_token_idx=2, trg_end_token_idx=3
+)  # set the arguments as per vocabulary index for '<' and '>'
+
+schedule_cb = tf.keras.callbacks.LearningRateScheduler(scheduler)
+
+model = Transformer(
+    num_hid=200,
+    num_head=2,
+    num_ff=400,
+    trg_maxlen=max_target_len,
+    num_layers_enc=4,
+    num_layers_dec=1,
+    num_classes=34,
+)
+model.compile(optimizer="adam")
+
+history = model.fit(ds, callbacks=[display_cb, schedule_cb], epochs=1)
```

Comment:
We should be monitoring a validation dataset during training (the text display callback is not sufficient for monitoring, we should be also looking at the validation loss to check overfitting)

Reply:
Okay, overriding `test_step` for this.